[Docs] gpl documentation (#3702)
The-OpenROAD-Project/OpenROAD: src/gpl/README.md

# Global Placement

> OpenROAD performs global placement using the RePlAce algorithm, from the

Reviewer: See `drt` PR for the correct phrasing.

# Global Placement

The global placement module in OpenROAD (`gpl`) is based on the open-source

> RePlAce algorithm, from the

Reviewer: ```suggestion
RePlAce tool, from the
```

## Commands

```{note}
- Parameters in square brackets `[-param param]` are optional.
- Parameters without square brackets `-param2 param2` are required.
```

### Global Placement

> `-timing_driven` does a virtual `repair_design` to find slacks and

Reviewer: ```suggestion
When using the `-timing_driven` flag, `gpl` does a virtual `repair_design` to find slacks and
```

    [-timing_driven_net_reweight_overflow]
    [-timing_driven_net_weight_max]
    [-timing_driven_nets_percentage]
```

#### General Arguments

| Switch Name | Description |
| ----- | ----- |
| `-timing_driven` | Enable timing-driven mode. |

> | `-routability_driven` | Enable routability-driven mode. |

Reviewer: Let's add here a "See [link] for timing/routability specific arguments."

    [-force_cpu]
```

#### Options

> | Switch Name | Description |

Reviewer: This is a long table; let's break it into three tables. One for arguments independent of `timing/routability` driven and one for each `timing/routability` only arguments. Please also order the arguments in `replace.tcl` to reflect this grouping.

| ----- | ----- |
| `-timing_driven` | Enable timing-driven mode. |
| `-routability_driven` | Enable routability-driven mode. |
| `-skip_initial_place` | Skip the initial placement (Biconjugate gradient stabilized, or BiCGSTAB solving) before Nesterov placement. Initial placement improves HPWL by ~5% on large designs. Equivalent to `-initial_place_max_iter 0`. | 
| `-incremental` | Enable the incremental global placement. Users would need to tune other parameters (e.g., `init_density_penalty`) with pre-placed solutions. | 

> | `-bin_grid_count` | Set bin grid's counts. Default value is defined by internal heuristic. Allowed values are  `[64,128,256,512,..., int]`. |

Reviewer: ```suggestion
| `-bin_grid_count` | Set bin grid's counts. The internal heuristic defines the default value. Allowed values are integers `[64,128,256,512,...]`. |
```

> | `-density` | Set target density. Default value is 0.70. Allowed values are `[0-1, float]`. |

Reviewer: ```suggestion
| `-density` | Set target density. The default value is `0.7` (i.e., 70%). Allowed values are floats `[0, 1]`. |
```

Reviewer: Let's use this notation: "floats/integers/etc `[x, y]`" Remember to use `[ | ]` for inclusive and `( | )` for excluding the first/last number.

| `-init_density_penalty` | Set initial density penalty. Default value is 8e-5. Allowed values are `[1e-6 - 1e6, float]`. |
| `-init_wirelength_coef` | Set initial wirelength coefficient. Default value is 0.25. Allowed values are `[unlimited, float]`. |
| `-min_phi_coef` | Set `pcof_min` ($\mu_k$ Lower Bound). Default value is 0.95. Allowed values are `[0.95-1.05, float]`. |

> | `-max_phi_coef` | Set `pcof_max` ($\mu_k$ Upper Bound). Default value is 1.05. Allowed values are `[1.00-1.20, float]`. |

Reviewer: Developer options?

Author: I am thinking it can potentially be tuned using autotuner?

Reviewer: Maybe we should consider using `Advanced Options` instead of `Developer Options`. Then, we can include AutoTuner + Dev.

Author: Lets discuss this tomorrow with @dralabeing

> | `-overflow` | Set target overflow for termination condition. Default value is 0.1. Allowed values are `[0-1, float]`. |

Reviewer: ```suggestion
| `-overflow` | Set target overflow for termination condition. The default value is `0.1`. Allowed values are floats `[0, 1]`. |
```

> | `-initial_place_max_iter` | Set maximum iterations in initial place. Default value is 20. Allowed values are `[0-MAX_INT, int]`. |

Reviewer: ```suggestion
| `-initial_place_max_iter` | Set maximum iterations in the initial place. The default value is 20. Allowed values are integers `[0, MAX_INT]`. |
```

> | `-initial_place_max_fanout` | Set net escape condition in initial place when $fanout \geq initial\_place\_max\_fanout$. Default value is 200. Allowed values are `[1-MAX_INT, int]`. |
> | `-timing_driven_net_reweight_overflow` | Set overflow threshold for timing-driven net reweighting. Allowed values are `tcl list of [0-100, int]`. |
> | `-timing_driven_net_weight_max` | Set the multiplier for the most timing critical nets. Default value is 1.9. |
> | `-timing_driven_nets_percentage` | Set the percentage of nets that are reweighted in timing-driven mode. Default value is 10. Allowed values are `[0-100, float]`. |
> | `-verbose_level` | Set verbose level for RePlAce. Default value is 1. Allowed values are `[0-5, int]`. |
> | `-force_cpu` | Force to use the CPU solver even if the GPU is available. | 

Reviewer: ```suggestion
| `-initial_place_max_fanout` | Set net escape condition in initial place when $fanout \geq initial\_place\_max\_fanout$. The default value is 200. Allowed values are integers `[1, MAX_INT]`. |
| `-timing_driven_net_reweight_overflow` | Set overflow threshold for timing-driven net reweighting. Allowed value is a Tcl list of integers where each number is `[0, 100]`. |
| `-timing_driven_net_weight_max` | Set the multiplier for the most timing-critical nets. The default value is `1.9`. |
| `-timing_driven_nets_percentage` | Set the reweighted percentage of nets in timing-driven mode. The default value is 10. Allowed values are floats `[0, 100]`. |
| `-verbose_level` | Set verbose level for `gpl`. The default value is 1. Allowed values are integers `[0, 5]`. |
| `-force_cpu` | Force to use the CPU solver even if the GPU is available. |
```

### Useful developer functions

> If you are a developer, you might find these useful. More details can be found in the [source file](./src/replace.cpp) or the [swig file](./src/replace.i).

Reviewer: In `drt` PR these links are full `https...` which is correct? Please use the same format in all docs.

```tcl
# debugging global placement 
global_placement_debug -pause -update -inst -draw_bins -initial

# adds padding and gets global placement uniform target density
get_global_placement_uniform_density -pad_left -pad_right 
```

## Example scripts

Example scripts demonstrating how to run RePlAce on a sample design on `core01` as follows:

```shell
./test/core01.tcl
```

## Regression tests

> There are a set of regression tests in `/test`. For more information, refer to this [section](../../README.md#regression-tests).

Reviewer: ```suggestion
There are a set of regression tests in `./test`. For more information, refer to this [section](../../README.md#regression-tests).
```

    [-max_split_size max_split_size]
```

#### General Arguments

| Switch Name | Description |
| ----- | ----- |
| `-tray_weight` | Set the weighting factor for tray cost (recommended to be `[20.0, float]`). |
| `-timing_weight` | Set the weighting factor for timing-critical paths in (recommended to be `[1.0. float]`). |
| `-max_split_size` | The maximum size of a single pointset after running the pointset decomposition algorithm for runtime improvement (to not run pointset decomposition, set as `-1`). |

> # Example Scripts

Reviewer: ```suggestion
## Example Scripts
```
